src/build.c

```diff
@@ -1649,3 +1751,129 @@ int build_conservation_phi(colloids_info_t * cinfo, field_t * phi,
 
   return 0;
 }
+
+/*****************************************************************************
+ *
+ *  Calculate surface normal on an ellipsoid at a point posvector
+ *
+ ****************************************************************************/
+__host__ __device__ void surface_normal_spheroid(colloid_t * pc,const double * posvector, double * rb) {
+
+  surface_vector_spheroid(pc,posvector, rb, 0);
+  return;
+ }
+
+/*****************************************************************************
+ *
+ *  Calculate surface tangent on an ellipsoid at a point posvector
+ *
+ ****************************************************************************/
+__host__ __device__ void surface_tangent_spheroid(colloid_t * pc,const double * posvector, double * rb) {
+
+  surface_vector_spheroid(pc,posvector, rb, 1);
+  return;
+ }
+
+/*****************************************************************************
+ *
+ *  Calculate surface normal vector on an ellipsoid at a point posvector
+ *
+ *  Note
+ *  "Speheroid" has a != b and b = c (for oblate or prolate case).
+ *
+ *  See, e.g., S.R. Keller and T.Y.T. Wu, J. Fluid Mech. 80, 259--278 (1977).
+ *
+ *  CHECK Does this always produce a unit vector?
+ *  CHECK what happens in the oblate case?
+ *
+ ****************************************************************************/
+__host__ __device__ void surface_vector_spheroid(colloid_t * pc,
+						 const double * posvector,
+						 double * rb,
+						 const int tn) {
+
+  double *elabc;
+  double elc;
+  double ele,ele2;
+  double ela,ela2;
+  double elz,elz2;
+  double elr;
+  double rmod;
+  double *elbz;
+  double denom, term1, term2;
+  double elrho[3];
+  double diff1,diff2,gridin[3],elzin,dr[3];
+
+  elabc=pc->s.elabc;
+  elc=sqrt(elabc[0]*elabc[0]-elabc[1]*elabc[1]);
+  ele=elc/elabc[0];
+  ela = colloids_largest_dimension(pc);
+  elbz=pc->s.m;
+  elz=dot_product(posvector,elbz);
+  for(int ia=0; ia<3; ia++) {elrho[ia]=posvector[ia]-elz*elbz[ia];}
+  elr = modulus(elrho);
+  rmod = 0.0;
+  if (elr != 0.0) rmod = 1.0/elr;
+  for(int ia=0; ia<3; ia++) {elrho[ia]=elrho[ia]*rmod;}
+  ela2=ela*ela;
+  elz2=elz*elz;
+  ele2=ele*ele;
+  diff1=ela2-elz2;
+  diff2=ela2-ele2*elz2;
+  /*Taking care of the unusual circumstances in which the grid point lies*/
+  /*outside the particle and elz > ela. Then the tangent vector is calculated*/
+  /*for the neighbouring grid point inside*/
+  if(diff1<0.0){
+    elr = modulus(posvector);
+    rmod = 0.0;
+    if (elr != 0.0) rmod = 1.0/elr;
+    for(int ia=0; ia<3; ia++) {dr[ia]=posvector[ia]*rmod;}
+    for(int ia = 0; ia < 3; ia++) {
```

Comment:
This looks a bit odd; you are computing the do product before you have set all the components of `gridin`.

Reply:
I think this is benign as it just executes the same thing 3 times... I move the offending lines